### PR TITLE
[GraphBolt] support DictItemSet in MinibatchSampler

### PR DESCRIPTION
**Pull request by Rhett-Ying**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/minibatch_sampler.py [hunk 1]
```diff
@@ -108,11 +100,72 @@ class MinibatchSampler(IterDataPipe):
...
+    >>> ids = {
+    ...     ("user", "like", "item"): gb.ItemSet(torch.arange(0, 5)),
+    ...     ("user", "follow", "user"): gb.ItemSet(torch.arange(5, 10)),
+    ... }
+    >>> item_set = gb.DictItemSet(ids)
+    >>> minibatch_sampler = gb.MinibatchSampler(item_set, 4)
+    >>> list(minibatch_sampler)
+    [{('user', 'like', 'item'): tensor([0, 1, 2, 3])},
+    {('user', 'follow', 'user'): tensor([5, 6, 7]),
+        ('user', 'like', 'item'): tensor([4])},
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch_sampler.py:114]: Shouldn't it be like this?
```suggestion
    {('user', 'like', 'item'): tensor([4])},
        ('user', 'follow', 'user'): tensor([5, 6, 7]),
```

File: python/dgl/graphbolt/minibatch_sampler.py [hunk 2]
```diff
@@ -108,11 +100,72 @@ class MinibatchSampler(IterDataPipe):
     >>> data_pipe = data_pipe.map(add_one)
     >>> list(data_pipe)
     [tensor([1, 2, 3, 4]), tensor([5, 6, 7, 8]), tensor([ 9, 10])]
+
+    7. Heterogeneous node/edge IDs.
+    >>> ids = {
+    ...     ("user", "like", "item"): gb.ItemSet(torch.arange(0, 5)),
+    ...     ("user", "follow", "user"): gb.ItemSet(torch.arange(5, 10)),
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch_sampler.py:107]: When we ask users to provide dataset for heterogeneous training.

For different node types, do we ask users to provide ids always from 0, or globally continuous?

File: python/dgl/graphbolt/minibatch_sampler.py [hunk 3]
```diff
@@ -108,11 +100,72 @@ class MinibatchSampler(IterDataPipe):
     >>> data_pipe = data_pipe.map(add_one)
     >>> list(data_pipe)
     [tensor([1, 2, 3, 4]), tensor([5, 6, 7, 8]), tensor([ 9, 10])]
+
+    7. Heterogeneous node/edge IDs.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch_sampler.py:104]: Heterogeneous node IDs.

edge IDs is not recommended in our new design, let's use node ID in example.

File: python/dgl/graphbolt/minibatch_sampler.py (hunk 2 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/minibatch_sampler.py:107]: I don't think we have any assumption or limitation on node ids. ids always could be arbitrary.

File: python/dgl/graphbolt/minibatch_sampler.py (hunk 1 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/minibatch_sampler.py:114]: No. `('user', 'follow', 'user')` is iterated and placed to the mini-batch first, then `('user', 'like', 'item')`. But anyway, `dict` is not ordered.

File: python/dgl/graphbolt/minibatch_sampler.py (hunk 3 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/minibatch_sampler.py:104]: what about homogeneous edge IDs? not recommended either?

Comment (frozenbugs) [reply python/dgl/graphbolt/minibatch_sampler.py:104]: No, not recommended either.

File: python/dgl/graphbolt/minibatch_sampler.py (hunk 2 above)
Comment (frozenbugs) [reply python/dgl/graphbolt/minibatch_sampler.py:107]: The itemset should not have any assumption, but as example, it should show what's we recommend to users.

File: python/dgl/graphbolt/minibatch_sampler.py (hunk 3 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/minibatch_sampler.py:104]: why eid is not recommended in graphbolt?

File: python/dgl/graphbolt/minibatch_sampler.py [hunk 4]
```diff
@@ -108,11 +100,71 @@ class MinibatchSampler(IterDataPipe):
...
+    >>> ids = {
+    ...     "user": gb.ItemSet(torch.arange(0, 5)),
+    ...     "item": gb.ItemSet(torch.arange(0, 5)),
+    ... }
+    >>> item_set = gb.DictItemSet(ids)
+    >>> minibatch_sampler = gb.MinibatchSampler(item_set, 4)
+    >>> list(minibatch_sampler)
+    [{'user': tensor([0, 1, 2, 3])},
+    {'item': tensor([0, 1, 2]), 'user': tensor([4])},
+    {'item': tensor([3, 4])}]
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch_sampler.py:114]:
```suggestion
    {'item': tensor([3, 4, 5])}]
```

File: python/dgl/graphbolt/minibatch_sampler.py [hunk 5]
```diff
@@ -108,11 +100,71 @@ class MinibatchSampler(IterDataPipe):
...
+    {('user', 'follow', 'user'): [tensor([0, 1, 2]), tensor([5, 6, 7])],
+        ('user', 'like', 'item'): [tensor([4]), tensor([9])]},
+    {('user', 'follow', 'user'): [tensor([3, 4]), tensor([8, 9])]}]
+
+    9. Heterogeneous node pairs and labels.
+    >>> node_pairs_labels = (
+    ...     torch.arange(0, 5), torch.arange(5, 10), torch.arange(10, 15))
+    >>> item_set = gb.DictItemSet({
+    ...     ("user", "like", "item"): gb.ItemSet(node_pairs_labels),
+    ...     ("user", "follow", "user"): gb.ItemSet(node_pairs_labels),
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch_sampler.py:134]: ditto

File: python/dgl/graphbolt/minibatch_sampler.py [hunk 6]
```diff
@@ -108,11 +100,71 @@ class MinibatchSampler(IterDataPipe):
...
+    {('user', 'follow', 'user'):
+        [tensor([3, 4]), tensor([8, 9]), tensor([13, 14])]}]
+
+    10. Heterogeneous head, tail and negative tails.
+    >>> head_tail_neg_tails = (
+    ...     torch.arange(0, 5), torch.arange(5, 10),
+    ...     torch.arange(10, 20).reshape(-1, 2))
+    >>> item_set = gb.DictItemSet({
+    ...     ("user", "like", "item"): gb.ItemSet(head_tail_neg_tails),
+    ...     ("user", "follow", "user"): gb.ItemSet(head_tail_neg_tails),
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch_sampler.py:152]: ditto.

**Review by frozenbugs — APPROVED** [review]

**Comment by jermainewang** [comment]
Some questions about heterogeneous item set.

1. Since the input is a dictionary, how will we decide the order when `shuffle=False`? For example, given `{ 'user' : ..., 'item' : ...}`, which of the "user" or "item" ids will be iterated first?
2. In the above example, if all the "user" ids have been consumed, will the returned mini-batch dictionary still contain the key `'user'` or not? If yes, what is shape/dtype of the corresponding tensor?

**Comment by Rhett-Ying** [comment]
> Some questions about heterogeneous item set.
> 
>     1. Since the input is a dictionary, how will we decide the order when `shuffle=False`? For example, given `{ 'user' : ..., 'item' : ...}`, which of the "user" or "item" ids will be iterated first?
> 
>     2. In the above example, if all the "user" ids have been consumed, will the returned mini-batch dictionary still contain the key `'user'` or not? If yes, what is shape/dtype of the corresponding tensor?

For 1, A number of items(this is controlled by buffer size) will be iterated and saved into a buffer, then pick one of them from the buffer when `shuffle=True` for each iteration. So in the case you mentioned, `user` will be iterated first if `len('user')` is greater than `buffer_size`(10000 in default). If not greater than `buffer_size`, then `item` may be picked up for each iteration. Here's the implementation which could help understand the shuffle logic: https://github.com/pytorch/pytorch/blob/6ac3352a37ead6bae0de9fe4c45b618f8e82dc78/torch/utils/data/datapipes/iter/combinatorics.py#L122-L135
So an enhancement is to dynamically specify the `buffer_size` to control the iteration order.

For 2, it only returns the ones picked up for the current mini-batch, so `No` if `user` has already been ran out.

@jermainewang

File: python/dgl/graphbolt/minibatch_sampler.py [hunk 7]
```diff
@@ -108,11 +100,71 @@ class MinibatchSampler(IterDataPipe):
...
+    >>> list(minibatch_sampler)
+    [{'user': tensor([0, 1, 2, 3])},
+    {'item': tensor([0, 1, 2]), 'user': tensor([4])},
+    {'item': tensor([3, 4])}]
+
+    8. Heterogeneous node pairs.
+    >>> node_pairs = (torch.arange(0, 5), torch.arange(5, 10))
+    >>> item_set = gb.DictItemSet({
+    ...     ("user", "like", "item"): gb.ItemSet(node_pairs),
+    ...     ("user", "follow", "user"): gb.ItemSet(node_pairs),
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply python/dgl/graphbolt/minibatch_sampler.py]: I changed to 2 different node_pairs

File: python/dgl/graphbolt/minibatch_sampler.py [hunk 8]
```diff
@@ -1,27 +1,19 @@
...
+from functools import partial
+from typing import Optional
 
 from torch.utils.data import default_collate
 from torchdata.datapipes.iter import IterableWrapper, IterDataPipe
 
 from ..batch import batch as dgl_batch
 from ..heterograph import DGLGraph
-from .itemset import ItemSet
+from .itemset import DictItemSet, ItemSet
```
Comment (jermainewang) [inline python/dgl/graphbolt/minibatch_sampler.py:12]: Small suggestion: perhaps rename `DictItemSet` to `ItemSetDict` because the former is not a subclass of `ItemSet`.

**Review by jermainewang — APPROVED** [review]

File: python/dgl/graphbolt/minibatch_sampler.py (hunk 8 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/minibatch_sampler.py:12]: pls let me change it in another PR to be clear.
